File: src/Core/Unmanaged/Portable/Lift.cpp
#include "pch.h"

#include <cassert>
#include "Lift.h"

using namespace LiftControl::Generation1;

vector<int> Lift::get_schedule()
{
    vector<int> result;
    if (NULL == this->m_direction)
    {
        return result;
    }
    else
    {
        assert(nullptr != this->m_pass1);
        assert(nullptr != this->m_pass2);
        assert(nullptr != this->m_pass3);
    }

    auto pass = this->m_pass1;
    LoopStart:
    for (auto floor : *pass)
    {
        result.push_back(floor);
    }
    if (pass == this->m_pass1)
    {
        pass = this->m_pass2;
        goto LoopStart;
    }
    else if (pass == this->m_pass2)
    {
        pass = this->m_pass3;
        goto LoopStart;
    }
    else
    {
        assert(pass = this->m_pass3);
    }

    return result;
}

void Lift::add(int from_floor, int to_floor)
{
	if (NULL == this->m_direction)
	{
		this->initialize(from_floor, to_floor);
	}

	auto guestDirection = GetDirection(from_floor, to_floor);
    if (this->m_direction == guestDirection)
    {
        if (this->m_atFloor == from_floor || this->m_direction == GetDirection(this->m_atFloor, from_floor))
        {
            this->m_pass1->insert(from_floor);
            this->m_pass1->insert(to_floor);
        }
        else
        {
            assert(this->m_direction == GetDirection(from_floor, this->m_atFloor));
            this->m_pass3->insert(from_floor);
            this->m_pass3->insert(to_floor);
        }
    }
    else
    {
        assert(this->m_direction != guestDirection);
        this->m_pass2->insert(from_floor);
        this->m_pass2->insert(to_floor);
    }
}

void Lift::initialize(int from_floor, int to_floor)
{
	auto guest_direction = GetDirection(from_floor, to_floor);
	this->m_direction = this->m_atFloor == from_floor ? guest_direction : GetDirection(this->m_atFloor, from_floor);
	FloorComparer comparer(this->m_direction), reversed_comparer = comparer.reverse();
	this->m_pass1 = new set<int, FloorComparer>(comparer);
	this->m_pass2 = new set<int, FloorComparer>(reversed_comparer);
	this->m_pass3 = new set<int, FloorComparer>(comparer);
}

Lift::~Lift()
{
    delete this->m_pass1;
    delete this->m_pass2;
    delete this->m_pass3;
}

File: src/Core/Managed/Portable/Generation1/Lift.cs
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;

namespace LiftControl.Generation1;

public class Lift(int atFloor = 0)
{
    protected int _atFloor = atFloor;
    protected Direction _direction;
    protected Queue<SortedSet<int>>? _schedule;

#if DEBUG
    public int AtFloor => this._atFloor;

    public IEnumerable<int> GetSchedule()
    {
        if (this._direction == Direction.None) yield break;

        this.CheckStatus();

        foreach (var pass in this._schedule)
        {
            foreach (var floor in pass)
            {
                yield return floor;
            }
        }
    }
#endif

    [MemberNotNull(nameof(_direction), nameof(_schedule))]
    protected void CheckStatus()
    {
        Debug.Assert(this._direction != Direction.None);
        Debug.Assert(this._schedule is not null);
    }

    public virtual void Add(int fromFloor, int toFloor)
    {
        var guestDirection = LiftUtilities.GetDirection(fromFloor, toFloor);
        if (this._direction == Direction.None)
        {
            Initialize();
        }

        [MemberNotNull(nameof(_direction), nameof(_schedule))]
        void Initialize()
        {
            this._direction = this._atFloor == fromFloor ? guestDirection : LiftUtilities.GetDirection(this._atFloor, fromFloor);
            this._schedule = new(4);
            var comparer = FloorComparer.GetInstance(this._direction);
            var reversedComparer = comparer.Reverse();
            this._schedule.Enqueue(new(comparer));
            this._schedule.Enqueue(new(reversedComparer));
            this._schedule.Enqueue(new(comparer));
        }

        this.CheckStatus();

        if (this._direction == guestDirection)
        {
            if (this._atFloor == fromFloor || this._direction == LiftUtilities.GetDirection(this._atFloor, fromFloor))
            {
                this._schedule.ElementAt(0).Add(fromFloor);
                this._schedule.ElementAt(0).Add(toFloor);
            }
            else
            {
                Debug.Assert(this._direction == LiftUtilities.GetDirection(fromFloor, this._atFloor));
                this._schedule.ElementAt(2).Add(fromFloor);
                this._schedule.ElementAt(2).Add(toFloor);
            }
        }
        else
        {
            Debug.Assert(this._direction != guestDirection);
            this._schedule.ElementAt(1).Add(fromFloor);
            this._schedule.ElementAt(1).Add(toFloor);
        }
    }
}
